gnc/sub8_controller/nodes/pd_controller.cc

```diff
@@ -0,0 +1,218 @@
+  control_timer = nh.createTimer(ros::Duration(control_period), &PDController::control_loop, this);
+}
+
+void PDController::control_loop(const ros::TimerEvent &timer_event) {
+  // Translation
+  Eigen::Vector3d world_force;
+  Eigen::Vector3d body_force;
+  Eigen::Vector3d translation_error;
+  Eigen::Vector3d translation_error_gradient;
+  Eigen::Vector3d translation_error_direction;
+  Eigen::Vector3d translation_error_integral;
+
+  // Orientation
+  Eigen::AngleAxis<double> orientation_error;
+  double angle;
+  Eigen::Vector3d angle_axis;
+  Eigen::Vector3d orientation_error_gradient;
+  Eigen::Vector3d orientation_error_integral;
+  Eigen::Vector3d world_torque;
+  Eigen::Vector3d body_torque;
+
+  // Common
+  geometry_msgs::WrenchStamped wrench_stamped_msg;
+  Eigen::Matrix<double, 6, 1> body_wrench;
+
+  if (not(got_state and got_target)) {
+    // Wait until we have some data to work with
+    return;
+  }
+
+  // ******* Translation Control *******
+  translation_error = target_state.position - current_state.position;
+  translation_error_direction = translation_error;
+  translation_error_direction.normalize();
+
+  append_to_history(translation_error_history, translation_error, trans_history_length);
+
+  translation_error_gradient = estimate_derivative(translation_error_history);
+  translation_error_integral = estimate_integral(translation_error_history);
+
+  // Convert error to body frame
+  world_force = ((kp_trans * translation_error) + (kd_trans * translation_error_gradient) +
+                 (ki_trans * translation_error_integral));
+
+  // Feed-Forward vehicle weight (This drastically improves performance)
+  world_force += sub_mass * 9.81 * Eigen::Vector3d::UnitZ();
+  body_force = current_state.orientation * world_force;
+
+  // ******* Orientation Control *******
+  orientation_error = rotation_difference(current_state.orientation, target_state.orientation);
+
+  // Normalize [-pi, pi]
+  angle = orientation_error.angle();
+  if (angle > M_PI) {
+    angle -= 2.0 * M_PI;
+  }
+
+  angle_axis = angle * orientation_error.axis();
+
+  append_to_history(orientation_error_history, angle_axis, angle_history_length);
+  orientation_error_gradient = estimate_derivative(orientation_error_history);
+  orientation_error_integral = estimate_integral(orientation_error_history);
+
+  world_torque =
+      -((kp_angle * angle_axis) + (kd_angle * orientation_error_gradient) + (ki_angle * orientation_error_integral));
+  body_torque = current_state.orientation * world_torque;
+
+  // ******* Transmit Wrenches *******
+  body_wrench << body_force, body_torque;
+  tf::wrenchEigenToMsg(body_wrench, wrench_stamped_msg.wrench);
+  wrench_stamped_msg.header.frame_id = "/body";
+  wrench_stamped_msg.header.stamp = ros::Time::now();
+  wrench_pub.publish(wrench_stamped_msg);
+}
+
+void PDController::append_to_history(controller_deque &history, Eigen::Vector3d &datum, unsigned int length) {
```

Comment:
https://google.github.io/styleguide/cppguide.html#Type_Names 
Should change "controller_deque"